### ISSUE TITLE
Do we need to check is_leaf?

### ISSUE THREAD

**Issue by jtlee90** [NVIDIA-AI-IOT/torch2trt#82]
https://github.com/NVIDIA-AI-IOT/torch2trt/blob/master/torch2trt/torch2trt.py#L127-L132

We employ TensorRT for accelerating inference in most of case, so I think checking the `is_leaf` may prevent converting 'weight' tensor. When we do the inference, we do not use a gradient(sometime for visualization used) and if tensor.requires_grad == True, then the tensor doesn't have `is_leaf` so that caused Assert error(if they use detach() it is okay) during the converting.

So, I mean, &*it is mainly for accelerating inference whatever the tensor is leaf or not, nobody will care**

I believe that these day a lot of research use self-attention mechanisms which use learnable weight parameter, like `input * weight`, then if with this code, it fail to generate trt_

What do you think? Do we have to check `is_leaf`, and what for?

**Comment by jtlee90** [comment]
jaybdub's mention is here.
> FYI, The main reason we check is_leaf is to make sure we don't create a constant for a tensor that actually isn't constant. This can happen when there is a PyTorch operation that isn't yet supported/registered in torch2trt. When a layer isn't supported, the failure doesn't actually occur until we later try to access the _trt, which is missing.

My concern is that, if we always produce constants, this failure may not happen, and the conversion may succeed and silently produce an incorrect model.

**Comment by jaybdub** [comment]
Hi jtlee90,

I've actually removed ``leaf_check`` as you suggested.  The methodology seemed not reliable when ``torch.no_grad`` was set.  This means that there can be silent failures (successful conversion, incorrect model).  However it seems``torch.jit`` has similar expectations

https://pytorch.org/docs/stable/jit.html#torch.jit.trace

Perhaps there is more that can be done to prevent this.  One simple (yet laborious) thing I can think of is manually list all PyTorch ops that modify tensors.  

```python
@tensorrt_converter('torch.unimplemented_op1')
@tensorrt_converter('torch.unimplemented_op2')
def convert_unimplemented(ctx):
    raise NotImplementedError
```

Another simple thing would be a sanity check at the end of conversion, that just cross-checks the final output against Python, and warns if a tolerance is exceeded.  It seems ``torch.jit`` also does this.

Please let me know if you have any suggestions.

Best,
John
